[MRG+1] Return nan in RadiusNeighborsRegressor for empty neighbor set

abjer opened the pull request:
#### Reference Issue
Fixes #9654 



#### What does this implement/fix? Explain your changes.
RadiusNeighborsRegressor is behaving differently when there are no neighbors for a sample between when weights are or aren't used. This PR fixes this inconsistency. This PR also fixes raised error when no available data points for `RadiusNeighborRegression` using non-uniform weights.

abjer committed:
Fix to issue 9654

Fixes problem with raise error when no available data points for RadiusNeighborRegression using non-uniform weights.

jnothman commented on sklearn/neighbors/regression.py:
```diff
@@ -291,9 +291,28 @@ def predict(self, X):
             y_pred = np.array([np.mean(_y[ind, :], axis=0)
```
Do we want this to be issuing a warning in this case? I think the tests should assert either that a warning is raised, or that none is.

abjer replied:
I have made a test now that asserts *no* warning is raised.

jnothman commented on sklearn/neighbors/regression.py:
```diff
@@ -291,9 +291,28 @@ def predict(self, X):
+
+            if np.min(has_neighors):
+                y_pred = np.array([(np.average(_y[ind, :], axis=0,
+                                               weights=weights[i]))
```
can't we just add `if len(ind) else np.nan` here?

jnothman replied:
Sorry, wrong line...

jnothman commented on sklearn/neighbors/regression.py:
```diff
@@ -291,9 +291,28 @@ def predict(self, X):
                                for ind in neigh_ind])
         else:
-            y_pred = np.array([(np.average(_y[ind, :], axis=0,
-                                           weights=weights[i]))
```
can't we just add `if len(ind) else np.nan` here?

abjer replied:
Right, great idea - I've changed this now.

abjer replied:
Hmm, on second thought I guess that it should match the second axis of _y? So inserting np.full(y_.shape[1], np.nan) under the else clause.

jnothman replied:
Yes, that's right. Is it tested?

abjer replied:
Yes, I've tested it - works for one or more _y variables. Will push it now.

abjer committed:
Simplify fix

Used @jnothman suggestion to simplify fix

abjer committed:
Attempt to fix test and nan dimensionality

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -658,6 +658,18 @@ def test_radius_neighbors_regressor(n_samples=40,
+                y_pred_alt_isnan = np.all(np.isnan(y_pred_alt))
+                raise_zero_div_error = False
+
+            except ZeroDivisionError:
```
This is unnecessary. The test will fail if there is an error raised.

I was asking about handling the warning which numpy issues when performing `np.mean([])`. I'm not sure whether it's good or bad for this warning to be issued, but we should be consistent in the weighted case. warnings are tested with our `assert_warns_message` and `assert_no_warnings` helpers.

abjer replied:
Right, good point. 

I looked into raising the warning - apparently warnings only arise when there are uniform weights. What  do you think of the following solution?

```python
# test fix for issue #9654
# test that nan is returned when no nearby observations        
X_test_nan = np.ones((1,n_features))*-1
if weights=='uniform':        
    assert_warns_message(RuntimeWarning, "Mean of empty slice.", 
                         neigh.predict, X_test_nan)

assert_true(np.all(np.isnan(neigh.predict(X_test_nan))))
```

abjer committed:
Updated test

abjer committed:
Bug fix in test: add import

abjer committed:
Fix pep8 requirements

abjer commented:
@jnothman I think it is ready now for merge - do you agree?

jnothman commented:
it's on my heap for review, but my throughput will be quite limited for the
next month or so.

On 8 Sep 2017 6:40 pm, "Andreas Bjerre-Nielsen" <notifications@github.com>
wrote:

> @jnothman <https://github.com/jnothman> I think it is ready now for merge
> - do you agree?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9655#issuecomment-328041240>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz661fnObwAevsSclMLbDafhHTHWsNks5sgP12gaJpZM4PH82n>
> .
>

abjer commented:
Alright, cheers!

jnothman commented:
> @jnothman I think it is ready now for merge - do you agree?


You should change the title from WIP to MRG

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -658,6 +659,15 @@ def test_radius_neighbors_regressor(n_samples=40,
+            # test that nan is returned when no nearby observations
+            X_test_nan = np.ones((1, n_features))*-1
+            if weights == 'uniform':
+                assert_warns_message(RuntimeWarning, "Mean of empty slice.",
```
No, we should issue a warning in all casess, not dependent on weights.

jnothman replied:
You should also get the output, i.e. `pred = assert_warns_message(...)` so that you can then do the isnan assertion.

abjer replied:
Hi again, thanks for the feedback! Just to be clear on what you want me to do: raise a warning for RadiusNeighborRegressor when it returns one or more nan values irrespective of using weights or not. I guess this should be consistent with and without weights. So should we copy the warning from np.mean (without weights) and issue it when y_pred contains nan rows (and there are weights)?

jnothman replied:
I suppose so. Either that, or we can be kinder to the user: hide the numpy error, and raise our own which is more explicit, with a message like "Some samples have no neighbors; predicting NaN."

abjer committed:
Empty warning for RadiusNeighborRegression

Added new warning when prediction of RadiusNeighborRegression returns empty values. Also suppressed warning for numpy warnings.

Moreover, added copyright and fixed erroneous description of output.

abjer committed:
Fix PEP8 issues

jnothman reviewed (commented):
Otherwise LGTM, thanks!

Please add an entry to whats_new

abjer committed:
Notation fix

Fixed more PEP8 and used @jnothman idea of using full_like

abjer committed:
More notation fix

jnothman commented on sklearn/neighbors/regression.py:
```diff
@@ -288,13 +292,29 @@ def predict(self, X):
+
+            with warnings.catch_warnings():
+                warnings.filterwarnings('ignore', "Mean of empty slice")
+                warnings.filterwarnings('ignore', ( "invalid value "
```
how about
```
                warnings.filterwarnings('ignore',
                                        "invalid value encountered in divide")
```

jnothman replied:
I wrote this before you fixed it.

jnothman commented:
LGTM

abjer commented:
Great :)

jnothman commented:
Please still add an entry in whats_new

abjer committed:
Update whats_new.rst: solution to issue #9654

jnothman commented on doc/whats_new.rst:
```diff
@@ -63,6 +63,17 @@ Decomposition, manifold learning and clustering
+  :class:`neighbors.RadiusNeighborsRegressor` can handle empty neighbor set 
+  when using non uniform weights. Also raises a new warning, irrespective of 
+  weighting, when no neighbors are found for one or more samples. Moreover, 
+  all numpy errors related to missing neighbors are suppressed. :issue:`9654`. 
```
We actually prefer the pull request number here, #9655.

jnothman commented on doc/whats_new.rst:
```diff
@@ -63,6 +63,17 @@ Decomposition, manifold learning and clustering
+- Fix bug so that the method ``predict`` for 
+  :class:`neighbors.RadiusNeighborsRegressor` can handle empty neighbor set 
+  when using non uniform weights. Also raises a new warning, irrespective of 
+  weighting, when no neighbors are found for one or more samples. Moreover, 
```
maybe just "... samples, in place of unclear numpy errors."

abjer replied:
Good suggestion, thanks

jnothman reviewed (commented):
Thanks again!

abjer committed:
Update whats_new.rst

abjer commented:
@jnothman do you need me to do anything more or is this ready now as it is?

jnothman commented:
Waiting for a second review

jnothman commented:
Please move your whats_new entry to doc/whats_new/v0.20.rst

abjer committed:
evert "Update whats_new.rst"

This reverts commit 8decafa5d2a5ee80be81f1e74ddfd01dbad908fa.

abjer committed:
Revert "Update whats_new.rst: solution to issue #9654"

This reverts commit 780db9ddef5bdd11ceb75d5c7d6fefd935e94cf7.

abjer committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn

abjer committed:
Updated whats new

abjer commented:
@jnothman: good suggestion - I've pushed an update

abjer commented:
@jnothman hi again joel, is there anything I can do to get a 2nd review? do you have some candidates for whom I should tag?

jnothman commented:
This is pretty straightforward, though a more thorough PR description might help a reviewer come on board. Key idea: `RadiusNeighborsRegressor` is behaving differently when there are no neighbors for a sample between when weights are or aren't used. Perhaps @qinhanmin2014, @lesteve...

abjer commented:
Thanks a lot Joel - will follow those suggestions

abjer commented:
Hi @qinhanmin2014  and @lesteve would you mind taking a look at this PR?

qinhanmin2014 commented on sklearn/neighbors/regression.py:
```diff
@@ -288,13 +292,28 @@ def predict(self, X):
+            with warnings.catch_warnings():
+                warnings.filterwarnings('ignore', "Mean of empty slice")
+                warn_div = "invalid value encountered in true_divide"
+                warnings.filterwarnings('ignore', warn_div)
```
(1)I don't think our code should change the default behaviour of other libraries. In this case, after user run RadiusNeighborsRegressor, he will not get any warning when he execute `np.mean([])`, it might be better to at least reset the warnings.
(2) `invalid value encountered in true_divide` seems a bit magical, could you please add some comments? Also, is there any specific reason you are blocking warnings here? It seems that you can use the same way as in the else clause. i.e.
```python
empty_obs = np.full_like(_y[0], np.nan)
y_pred = np.array([np.mean(_y[ind, :], axis=0, weights=weights[i]) if len(ind) else empty_obs
                             for (i, ind) in enumerate(neigh_ind)])
```

jnothman replied:
The with catch_warnings context will reset when exited. But is there a reason we are not using numpy's own error manager?

jnothman replied:
The with catch_warnings context will reset when exited. But is there a reason we are not using numpy's own error manager?

abjer replied:
No particular reason. I can implement the numpy warnings if you wish it to be changed.

qinhanmin2014 replied:
@jnothman Thanks a lot for the confirm. If you think blocking warnings here is the right way, I won't oppose. My main concern is the redundant tests for this edge case
@abjer If you don't have specific reason, please consider to address @jnothman's comment here. Also, please consider whether we should cut off some redundant tests.
Another thing @abjer could you please update the document to tell users what we are doing in this special case? I think it might not be so straightforward for ordinary users.

qinhanmin2014 replied:
@abjer You need to solve the conflict before this PR can be merged (e.g., merge current master into this branch). Thanks a lot for your issue and PR and also for your great patience :)

qinhanmin2014 commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -658,6 +659,17 @@ def test_radius_neighbors_regressor(n_samples=40,
             assert_true(np.all(abs(y_pred - y_target) < radius / 2))
 
+            # test fix for issue #9654
+            # test that nan is returned when no nearby observations
```
If you put the test here, similar test will execute 12 times(4 algorithm X 3 weight), but seems that the change is actually unrelated with algorithm? I think a seperate test(e.g., test_radius_neighbors_regressor_nan_output) with 2 cases (uniform weight and distance weight) might be enough.

qinhanmin2014 reviewed (commented):
Still some concerns before my personal LGTM. Kindly forgive if some of them are naive.
@jnothman would be grateful if you can check these comments.

abjer committed:
Updated fct. and tests

abjer committed:
Fix typo

abjer committed:
bug fix: re-introduce empty obs

qinhanmin2014 commented on sklearn/neighbors/regression.py:
```diff
@@ -292,15 +296,29 @@ def predict(self, X):
-            y_pred = np.array([np.mean(_y[ind, :], axis=0)
-                               for ind in neigh_ind])
+            y_pred = np.array([np.mean(_y[ind, :],
+                                       axis=0)
```
please consider to fill the current line before starting a new line, other places the same.

qinhanmin2014 commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -658,6 +659,29 @@ def test_radius_neighbors_regressor(n_samples=40,
+
+        empty_warning_msg = ("One or more samples have no neighbors "
+                             "within specified radius; predicting NaN.")
+
```
please consider to remove some unnecessary blank lines both in the code and in the test.

qinhanmin2014 commented on doc/whats_new/v0.20.rst:
```diff
@@ -179,6 +179,14 @@ Metrics
 
+  Neighbors
+
+  - Fix bug so that the method ``predict`` for
```
Fixed a bug seems more consistent?

qinhanmin2014 commented on doc/whats_new/v0.20.rst:
```diff
@@ -179,6 +179,14 @@ Metrics
+    :class:`neighbors.RadiusNeighborsRegressor` can handle empty neighbor set
+    when using non uniform weights. Also raises a new warning, irrespective of
+    weighting, when no neighbors are found for samples, in place of unclear
+    numpy errors.. :issue:`9655`. By :user:`Andreas Bjerre-Nielsen <abjer>`.
```
duplicate . (numpy errors..)

qinhanmin2014 commented on doc/whats_new/v0.20.rst:
```diff
@@ -179,6 +179,14 @@ Metrics
   :issue:`10093` by :user:`alexryndin <alexryndin>`
   and :user:`Hanmin Qin <qinhanmin2014>`.
 
+  Neighbors
```
The [doc](https://14971-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/whats_new/v0.20.html) does not look fine, please follow what others are doing (not sure but maybe remove some extra whitespace at the beginning?).

qinhanmin2014 reviewed (commented):
Could you please fix the flake8 errors according to Travis log (See the bottom of https://travis-ci.org/scikit-learn/scikit-learn/jobs/304461869)? I think my LGTM is very close :)

abjer committed:
Fix pep8 error and tighten notation

abjer committed:
erge branch 'fix_9654' of github.com:abjer/scikit-learn

abjer committed:
Fix more PEP8 bugs.

qinhanmin2014 commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -658,6 +659,23 @@ def test_radius_neighbors_regressor(n_samples=40,
+                                    X_test_nan)
+        assert_true(np.all(np.isnan(pred)))
+
+
```
@abjer Please only leave two blank lines between functions. This flake8 error will not be detected for some reason but we generally don't want to introduce more flake8 errors.

abjer replied:
Thanks, I have now recommitted.

abjer committed:
Fix PEP8 lines space between fct.

qinhanmin2014 reviewed (approved):
LGTM ping @jnothman 
Latest change since your +1
(1) not block warnings, use similar way as another case instead.
(2)remove redundant tests, only keep two tests for the two cases.

abjer commented:
@qinhanmin2014 thank you so much for the review and very constructive suggestions

jnothman commented:
Great!

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -179,6 +179,13 @@ Metrics
+
+- Fixed a bug so ``predict`` in :class:`neighbors.RadiusNeighborsRegressor` can
+  handle empty neighbor set when using non uniform weights. Also raises a new
+  warning, irrespective of, when no neighbors are found for samples, in place of
```
"Irrespective of" can be removed

jnothman replied:
I'm not sure what in place of duplicate means

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -179,6 +179,13 @@ Metrics
+- Fixed a bug so ``predict`` in :class:`neighbors.RadiusNeighborsRegressor` can
+  handle empty neighbor set when using non uniform weights. Also raises a new
+  warning, irrespective of, when no neighbors are found for samples, in place of
+  duplicate. (numpy errors..) :issue:`9655`. By :user:`Andreas Bjerre-Nielsen <abjer>`.
```
What is "numpy errors" here. Drop it

jnothman reviewed (commented):
Apart from what's new being unclear, this is good to merge

qinhanmin2014 commented:
> Apart from what's new being unclear, this is good to merge

Sorry that I somehow miss the what's new in the final check

@abjer I believe current what's new is just a misoperation? The previous what's new seems fine (except for the format I've pointed out). So please correct current what's new accordingly. I'll merge when CIs are green.

abjer committed:
Make whats new more clear

abjer committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn

abjer commented:
@jnothman thanks for pointing that out - I have submitted a new version

qinhanmin2014 commented:
Thanks @abjer for your contributions and your patience :)
